README.md
<a href="https://hits.seeyoufarm.com"><img src="https://hits.seeyoufarm.com/api/count/incr/badge.svg?url=https%3A%2F%2Fgithub.com%2Fmoshix%2Fzlinux&count_bg=%2379C83D&title_bg=%23555555&icon=&icon_color=%23E7E7E7&title=hits&edge_flat=false"/></a>

zLinux Installer
================


This is a collection of scripts, recipes and stratagems to obtain an Ubuntu 18.04 ISO image for the s390x architecture, and then install it on an IBM 3390 virtual disk under Hercules. It was put together by Matthew Wilson and moshix over a period of a week or so.  

:tada: Just launch the script, have yourself a nice cup of coffee, and return to a fully installed Ubuntu/s390x system!

Features
========

| Feature                    | Supported          |
| -------------------------- | ------------------ |
| Hands-off installation     | :white_check_mark: |  
| Automatic ISO download     | :white_check_mark: |  
| Automagic DASD creation    | :white_check_mark: |  
| Full network capability    | :white_check_mark: |  
| Automatic network routing  | :white_check_mark: |  
| Hercules binaries included | :white_check_mark: |  
| Windows host               | :x:                |  

  

Installation
============

Start the installer by doing:

>./zlinux_install.bash

The script will proceed to download Ubuntu 18.0.4-5 for s390x ISO to your local machine.  

It will then ask you to pick a password for the user "zubuntu" on the new system.

After that it will automatically start an up-to-date Hercules emulator (which is provided with this repo) and run the full Ubuntu 18.04 server installation procedure, fully automated. No user interaction is needed for the complete installaton.  

The installation can take 90 minutes to 2 hours, depending on the speed of your machine.  


A standard server Ubuntu 18.04 will be installed with an ssh server, so you will be able to ssh into Linux. 


Operation
=========

After the successful install, you start Ubuntu by running the following script:

>./run_zlinux.bash

Once Linux has finished booting, it has full connectivity to the Internet (thru NAT) and you can ssh into the user zubuntu by doing:

>ssh zubuntu@10.1.1.2

and provide the password you specified for this user during installation.

Everything else is just normal Linux; i.e. you can run apt-get, etc., etc.

A few things to know about the generated Ubuntu instance that will IPL from DASD after the install:

>Internal IP:   10.1.1.2  
>IP of gateway: 10.1.1.1  
>DNS server:    1.1.1.1  
>user name:     zubuntu/password as set up by you during installation process  

The s/390x Linux instance will connect through a tunnel interface which is set up by the the run_zlinux.bash script in this directory. It then further uses NAT on your default NIC interface to connect to the Internet.



Why Ubuntu 18.04
================

As of June 2023 and as of Hyperion Hercules version 4.6, Ubuntu 18.04 is the newest version that can be IPLd on Hercules. If you attempt to upgrade an existing Ubuntu 18.04 install to version 20.04 or higher, it will fail on IPL.  

By the same token, the last version of CentOS (or RHEL) that will IPL on the current Hercules is CentOS 7, or ClefOS 7.  

The reason for this is that later versions of Linux require special CPU instructions by the real iron mainframes of architecture level z12 and up, which Hyperion Hercules currently is not able to emulate. At the time of this writing, we don't expect a change in this situation any time soon. 


After Installation
==================

You will have Ubuntu/s390x installed in dasd/hd0.120

You can remove all unnecessary files by executing

>./cleanup_after_successful_install.bash



Racingmars/Moshix    
June, 2023  



Changelog
v 0.2
-----
use our own supplied Hercules version
Use scripts to get iso, create install directory, create initrd
Put hercules.cnf in assets and swap after finished install
Check for sudo execution of script or warn user
Ask user to run newly installed zLinux system
auto-tune hercules.cnf
run sanity checks for DASD size selection
check if iso was really downloaded before attempting to mount it
made 3GB vs 9GB more permissive in selection
added cleanup_after_successful_install.bash script to make cleaning up easy
if not wget then use curl
deal with file permissions
more sanity checks (for example see if install/ dir has content etc.)
don't get ISO if we have it already (somebody could rerun the procedure
multiple times)
don't allow user to be root


v 0.3
-----
Made tuning logic more streamlined
Same for RAM tuning
created run_zlinux script to run Ubuntu and setup NAT and tunneling
Expanded README
made dasd/ and logs/ directory non-empty to make it compatible with git
Made assets/hercules.rc.hd0 able to shutdown
created run_zlinux.bash script to set up environment and start zLinux
Fixed tuning for the nth+1 time...


v 0.4
-----
Removed option to start freshly installed Ubuntu from install script
create dasd/ and logs/ if they don't exist
fixed small stuff in run_zlinux.bash
Updated hercules.cnf for FACILITY instead of ARCHLVL
updated hercules.rc.hd0
fixed ipl.rc
export HERCULES_RC properly
auto-tune now finally works. It's been only 3 weeks fiddling with it...
set assets/ to non-writable


v 0.5 (most  by Racingmars)
-------------------------
General script cleanup and fixes
Fix RAM auto-tune
Offer 27GB (3390-27) disk type; improve DASD creation
Switch to template-based preseed and hercules.cnf creation. Only create config
during initial installation; don't overwrite user changes on future runs
Fix wget/curl detection; improvements to ISO download and mount
Use a Hercules built with ./herc4x/lib and ./herc4x/lib/hercules runpath;
include build script for reference
Run as regular user, not with sudo or as root. Run specific commands with sudo
Detect installation success
Only make one log file per script per run
Don't remove write permissions on anything -- we want user to be able to
easily rm this entire directory without trouble if they want to
Mark apparent success or failure of Ubuntu automated install process
Enahnced hercules console automation 
logs have proper time stamps in file name now
check if enough disk space exists before allocating DASD

v 0.6
-----
Check for net-tools availability; the set_network script needs it
Check if if the scripts are being executed from zlinux/ 
Fix README a bit
more logging, especially at the end of Ubuntu install automated procedure
Fixed space requirements calculations
fixed variable double quoting here and there
Other syntax improvements 
Automated shellcheck syntax checking before pushing repo...zlinux embraces
devops
net-tools no longer needed. We use it, if there, but not required
cleaned up syntax a bit more

v0.7
----

whole new set of Hyperion binaries, built for wider compatibiity with new CPUs
README updates
tested with more systems
Switched to sudo hercules launch for now
FIX: use sudo -E to keep path for newer distros


run_zlinux.bash
#!/usr/bin/env bash

# Copyright 2022 by moshix
# sets up networking etc. and then IPLs zlinux
# Uses a provided Hercules environment and not any pre-installed version of Hercules

# v0.1 copied over a lot of stuff from installer script
# v0.2 fixed sudo stuff
# v0.3 fixed RAM calculation
# v0.4 more cleanup; remove unused, unnecessary, and broken code;
#      always exit with error status when exiting due to error.
# v0.5 use the Hercules config as it was created at install time, or later
#      modified by the user. Do not recalc CPU and RAM tuning.
# v0.6 use a Hercules build with a relative rpath set
# v0.7 run as regular user, only using sudo when necessary
# v0.8 consistent log file name for all message during one run
# v0.9 check for installation success upton launching
# v1.0 Check if script is being executed from correct directory
# v1.1 More logging
# v1.2 Follow proper syntax 


source ./Version



# This is the command we will use when we need superuser privileges.
# If you use "doas" you may change it here.
SUDO="sudo -E"

test_sudo () {
    echo "${yellow}Testing if '$SUDO' command works ${reset}"
    if [[ $($SUDO id -u) -ne 0 ]]; then
        echo "${rev}${red}$SUDO did not set us to uid 0; you must run this script with a user that has $SUDO privileges.${reset}"
        exit 1
    fi
}

check_if_root () {
    # check if I am root and terminate if so
    if [[ $(id -u) -eq 0 ]]; then
        echo "${rev}${red}You are root. There is no need to be root to run zlinux. Please run as a normal user...${reset}"
        exit 1
    fi
}

logextension=`date "+%F-%T"`
logit () {
    # log to file all messages
    logdate=`date "+%F-%T"`
    echo "$logdate:$1" >> ./logs/zLinux_runtime.log."$logextension"
}

set_colors() {
    red=`tput setaf 1`
    green=`tput setaf 2`
    yellow=`tput setaf 3`
    blue=`tput setaf 4`
    magenta=`tput setaf 5`
    cyan=`tput setaf 6`
    white=`tput setaf 7`
    bold=`tput bold`
    uline=`tput smul`
    blink=`tput blink`
    rev=`tput rev`
    reset=`tput sgr0`
}

check_os () {
    if [[ "$OSTYPE" == "linux-gnu"* ]]; then
        echo
    elif [[ "$OSTYPE" == "darwin"* ]]; then
        echo "${red}MacOS detected. Sorry, MacOS is not yet supported.${reset}"
        exit 1
    elif [[ "$OSTYPE" == "cygwin" ]]; then
        echo "${red}Cygwin detected. Sorry, Cygwin is not supported.${reset}"
        exit 1
    elif [[ "$OSTYPE" == "win32" ]]; then
        echo "${red}Windows detected. Sorry, Windows is not supported.${reset}"
        exit 1
    else
        echo "${red}Unrecognzied operating system. Exiting now.${reset}"
        exit 1
    fi
}

check_install_success () {
    if [[ ! -f install_success ]]; then
        logit "No install_success file found"
        echo "${rev}${red}It does not appear an installation has succeeded."
        echo "Please run ./zlinus_install.bash to install your zlinux system.${reset}"
        exit 1
    fi
}

set_hercenv () {
    # set path to supplied hercules
    export PATH=./herc4x/bin:$PATH
}

run_sudo () {
    # are we running as sudo? If not inform user and exit
    arewesudo=`id -u`
    if [ "$arewesudo"  -ne 0 ]; then
        echo "${red}${rev}You need to execute this script with sudo or it won't work!${reset}"
        exit 1
    fi
}

# main starts here

set_colors

# are we running from the zlinux directory (instead of higher or lower directory?)
cpwd=`pwd`
curdir=`basename "$cpwd"`

if [[ "$curdir"  != "zlinux" ]]; then
   echo "${rev}${red}This script needs to be executed from inside the zlinux directory. Please retry. ${red}"
   exit 1
fi



check_if_root   # cannot be root
logit "user invoking run script: $(whoami)"

test_sudo       # but we must have sudo capability

# quick sanity checks
check_os
check_install_success

set_hercenv   # set paths for local herc4x hyperion instance

echo; echo; echo
logit "executing set_network script"

# execute network configurator
$SUDO ./scripts/set_network
$SUDO chown $(id -u):$(id -g) ./logs/setnetwork.log*

# The hercifc util needs to be setuid root to manage network interface. We will
# copy the as-installed copy, which is owned by the user and not root, so that
# we can restore the original owner and permissions after Hercules is done.
# But we won't overwrite an existing backup, in case previous runs failed and
# we're starting again
if [[ ! -f herc4x/bin/hercifc.orig ]]; then
    cp herc4x/bin/hercifc herc4x/bin/hercifc.orig
fi
logit "Starting zLinux "
$SUDO chown root:root herc4x/bin/hercifc
$SUDO chmod +s herc4x/bin/hercifc
echo "${bold}${uline}${yellow}Starting zLinux now....${reset}"
sleep 1.2

logdate=`date "+%F-%T"`
FILE=./logs/hercules.log.$logdate
HERCULES_RC=hercules.rc hercules -f hercules.cnf > "$FILE"

# After hercules finishes, restore the original hercifc
# This is for two reasons: less time leaving a suid binary sitting around,
# and make it easier for the user to clean up later by "rm -rf ..." this
# entire directory and not run into problems deleting a file owned by root.
$SUDO rm -f herc4x/bin/hercifc
mv herc4x/bin/hercifc.orig herc4x/bin/hercifc

logit "Ending zlinux"

# moshix LICENSES THE LICENSED SOFTWARE "AS IS," AND MAKES NO EXPRESS OR IMPLIED
# WARRANTY OF ANY KIND. moshix SPECIFICALLY DISCLAIMS ALL INDIRECT OR IMPLIED
# WARRANTIES TO THE FULL EXTENT ALLOWED BY APPLICABLE LAW, INCLUDING WITHOUT
# LIMITATION ALL IMPLIED WARRANTIES OF, NON-INFRINGEMENT, MERCHANTABILITY, TITLE
# OR FITNESS FOR ANY PARTICULAR PURPOSE. NO ORAL OR WRITTEN INFORMATION OR ADVICE
# GIVEN BY moshix, ITS AGENTS OR EMPLOYEES SHALL CREATE A WARRANTY


scripts/build-hyperion.sh
#!/bin/sh

set -ue

# THIS SCRIPT IS INCLUDED FOR REFERENCE AND DISTRIBUTION MAINTENANCE.
# THERE IS NO EXPECTATION ZLINUX/ZUBUNTU TURNKEY USERS WILL RUN IT.

# This script builds the latest SDL-Hyperion from git. The goal is to build a
# "relocatable" set of binaries and libraries so that we can distribute this in
# the turnkey zUbuntu system.
#
# To do this, we will make sure the build artifacts have a rpath set to a
# relative directory location, namely ./herc4x/lib. This will ensure that if
# you run the "hercules" command from a directory that has the ./herc4x/lib
# subdirectory, the dynamic linker will find the Hercules libraries.
#
# There is a key advantage to doing this over setting LD_LIBRARY_PATH at
# runtime: the hercifc binary will need to be setuid root, but setuid binaries
# ignore LD_LIBRARY_PATH. That won't work for our purposes, so we need ld.so to
# find the necessary libraries without any environment overrides.

# For the zlinux turnkey distribution, we run this script on Debian 9, an old
# enough distribution that the libc we link against should be compatible with a
# wide range of Linux distributions.

### IMPORTANT: this is really just a one-time script to run in a clean
### Debian build server. If you run multiple times on the same system,
### you're responsible for cleaning up before running it again.

SUDO="sudo"

$SUDO apt-get update -y
$SUDO apt-get install -y zlib1g-dev build-essential git \
	libltdl-dev libbz2-dev

git clone https://github.com/SDL-Hercules-390/hyperion.git \
	sdl-hyperion-git
mkdir sdl-hyperion-build
cd sdl-hyperion-build
../sdl-hyperion-git/configure --prefix=/tmp/herc4x \
	CFLAGS="-m64 -march=x86-64 -mtune=generic -O1" \
	LDFLAGS="-Wl,-rpath=./herc4x/lib,-rpath=./herc4x/lib/hercules" \
	LT_SYS_LIBRARY_PATH=./herc4x/lib:./herc4x/lib/hercules
make -j 2
make install
cd ..

# Include the Hercules copyright notice in the distributed binary release.
cp sdl-hyperion-git/COPYRIGHT /tmp/herc4x

tar -c -f herc4x.tgz -z -C /tmp herc4x

echo "*** herc4x distribution ready in herc4x.tgz"



cleanup_after_successful_install.bash
#!/usr/bin/env bash

# Copyright 2022 by moshix
# cleanup script after successful install
# removes the following:
#   install/ directory
#   logs/ files relating to installation
#   removes the *.iso files (can be several files!)

# v0.3 clean up
# v0.4 no sudo required
# v0.5 consistent log file name for all messages during one run
# v0.6 Check if script is being executed from correct directory

source ./Version

logextension=`date "+%F-%T"`
logit () {
    # log to file all messages
    logdate=`date "+%F-%T"`
    echo "$logdate:$1" >> ./logs/cleanup.log.$logextension
}

set_colors() {
    red=`tput setaf 1`
    green=`tput setaf 2`
    yellow=`tput setaf 3`
    blue=`tput setaf 4`
    magenta=`tput setaf 5`
    cyan=`tput setaf 6`
    white=`tput setaf 7`
    blink=`tput blink`
    rev=`tput rev`
    reset=`tput sgr0`
}

clean_stuff () {
    rm -fr install/
    rm -fr logs/install*
    rm -fr ./ubuntu-18.04.5-server-s390x.iso*

    echo "${yellow}All clean now! Goodbye. ${reset}"
    exit 0
}

# main starts here

set_colors

# are we running from the zlinux directory (instead of higher or lower directory?)
cpwd=`pwd`
curdir=`basename "$cpwd"`

if [[ "$curdir"  != "zlinux" ]]; then
  echo "${rev}${red}This script needs to be executed from inside the zlinux directory. Please retry. ${red}"
  exit 1
fi


echo "${yellow}Cleanup Procedure Version: $version${reset}"
echo  " "

sleep 1


while true; do
    read -p "${white} Do you want to clean up your directory after a successful install? (y/n) ${reset}" runvar

    case "$runvar" in
    [Yy]*)
        echo "${yellow}Roger, cleaning it all up now... ${reset}"
        clean_stuff
        ;;
    [Nn]*)
        echo "${yellow}Ok, terminating now....  ${reset}"
        exit
        ;;
    *)
        echo "${red}Unrecognized selection: $runvar. y or n  ${reset}" ;;
    esac
done


zlinux_install.bash
#!/usr/bin/env bash

# Copyright 2022 by moshix
# Installation script for S390x Ubuntu 18.04 .
# Does not necessarily work with prior or later version of Ubuntu.
# Obtains virgin ISO from Ubuntu, runs hercules-based install and then creates runtime environment.
# Uses a provided Hercules environment and not any pre-installed version of Hercules.

# v0.1 get location of ISO
# v0.2 inform user where it's being taken from
# v0.3 use scripts/ scripts to download, mount, create hdisk
# v0.4 Use our supplied hercules
# v0.5 detect OS and distro and block if not compatible
# v0.6 check for sudo
# v0.7 fix paths to relative not to userid relative (not ~./ )
# v0.8 auto-tune hercules.cnf
# v0.9 ask user to press enter to start hercules so they can see it will last a while
# v1.0 fix permissions
# v1.1 don't allow execution as user root
# v1.2 fixed RAM calculation
# v1.3 remove unused, unnecessary, and broken code;
#      always exist with error status when exiting due to error
# v1.4 do not leave directories without execute bit set
# v1.5 offer 27GB (3390-27) disk option; improve DASD creation
# v1.6 switch to template-based preseed and hercules.cnf creation
# v1.7 use a Hercules build with a relative rpath set
# v1.8 run as regular user, only using sudo when necessary
# v1.9 detect successful installation
# v1.10 misc fixes:
#         put all log messages for this run in same log file
#         fixup permissions of set_network log
#         dasdinit log typo
#         rename assets directory to templates
# v1.11 Make sure we have the route command available before we go too far
# v1.12 Syntax improvements, better logging
# v1.13 remove requirement for net-tools

source ./Version

# This is the command we will use when we need superuser privileges. It is
# exported so scripts we call will also use this value. If you use "doas" you
# may change it here.
SUDO="sudo -E"
export SUDO

test_sudo () {
#    echo "${yellow}Testing if '$SUDO' command works ${reset}"
    if [[ $($SUDO id -u) -ne 0 ]]; then
        echo "${rev}${red}$SUDO did not set us to uid 0; you must run this script with a user that has $SUDO privileges.${reset}"
        exit 1
    fi
}

check_if_root () {
    # check if I am root and terminate if so
    if [[ $(id -u) -eq 0 ]]; then
        echo "${rev}${red}You are root. You must run this installer as a regular user. Terminating...${reset}"
        exit 1
    fi
}

logextension=`date "+%F-%T"`
logit () {
    # log to file all messages
    logdate=`date "+%F-%T"`
    echo "$logdate:$1" >> ./logs/zLinux_installer.log."$logextension"
}

set_colors() {
    red=`tput setaf 1`
    green=`tput setaf 2`
    yellow=`tput setaf 3`
    blue=`tput setaf 4`
    magenta=`tput setaf 5`
    cyan=`tput setaf 6`
    white=`tput setaf 7`
    bold=`tput bold`
    uline=`tput smul`
    blink=`tput blink`
    rev=`tput rev`
    reset=`tput sgr0`
}

check_os () {
    if [[ "$OSTYPE" == "linux-gnu"* ]]; then
        echo
    elif [[ "$OSTYPE" == "darwin"* ]]; then
        echo "${rev}${red}MacOS detected. Sorry, MacOS is not yet supported.${reset}"
        exit 1
    elif [[ "$OSTYPE" == "cygwin" ]]; then
        echo "${rev}${red}Cygwin detected. Sorry, Cygwin is not supported.${reset}"
        exit 1
    elif [[ "$OSTYPE" == "win32" ]]; then
        echo "${rev}${red}Windows detected. Sorry, Windows is not supported.${reset}"
        exit 1
    else
        echo "${rev}${red}Unrecognized operating system. Exiting now.${reset}"
        exit 1
    fi
}

get_cores () {
    # tune CPU number
    cores=`grep -c ^processor /proc/cpuinfo`
    intcores=1

    if [[ $cores -gt 7 ]]; then
        intcores=6
    elif [[ $cores -gt 5 ]]; then
        intcores=5
    elif [[ $cores -gt 3 ]]; then
        intcores=2
    else
        intcores=1
    fi

    echo "${yellow}Number of cores present ${cyan} $cores. ${yellow}Setting Hercules to ${cyan} $intcores ${reset}"
    sleep 2
}

get_ram () {
    # this function sets a sensible amount of RAM for the Ubuntu/s390x installation procedure
    bkram=`grep MemTotal /proc/meminfo | awk '{print $2}'`
    let "gbram=$bkram/1024"

    if [[ $gbram -lt 1300 ]]; then
        echo "${rev}${red}You have only ${cyan} $gbram ${red} in RAM in your system. That is not enough to install zLinux. Exiting now. ${reset}"
        exit 1
    elif [[ $gbram -lt 2200 ]]; then
        hercram=1024
    elif [[ $gbram -lt 6000 ]]; then
        hercram=2048
    elif [[ $gbram -lt 16000 ]]; then
        hercram=4096
    else
        hercram=8192
    fi

    echo "${yellow}RAM in KB ${cyan}${gbram}. ${yellow}Setting Hercules to ${cyan}${hercram}${reset}"
    sleep 2
}

set_hercenv () {
    # set path to supplied hercules
    export PATH=./herc4x/bin:$PATH
}

create_conf () {
    cp templates/hercules.cnf.template hercules.cnf
    chmod +w hercules.cnf
    sed -i "s/__CPU__/$intcores/" hercules.cnf
    sed -i "s/__RAM__/$hercram/" hercules.cnf
}

check_already_installed () {
    if [[ -f install_success ]]; then
        logit "there is already an install_success file"
        echo "${yellow}It appears you have already completed a successful installation."
        echo -n "Do you really want to start over? (Y/N):${reset} "
        read startover
        [[ "$startover" == [yY] || "$startover" == [yY][eE][sS] ]] || exit 1
        logit "user chose to overwrite current installation"
        rm -f install_success
    fi
}

# main starts here

set_colors
# are we running from the zlinux directory (instead of higher or lower directory?)
cpwd=`pwd`
curdir=`basename "$cpwd"`

if [[ "$curdir"  != "zlinux" ]]; then
	echo "${rev}${red}This script needs to be executed from inside the zlinux directory. Please retry. ${red}"
	exit 1
fi


mkdir -p logs/
mkdir -p dasd/


check_if_root # cannot be root
logit "user invoking install script: $(whoami)"

check_already_installed

test_sudo     # but we must have sudo capability

# quick sanity checks
check_os

get_cores
get_ram
create_conf

set_hercenv

echo
echo
echo "${green}    Starting zLinux installer Version $version ${reset}"
echo "${green}    ===================================== ${reset}"
echo
logit "Starting zLinux installer"

# ask user for disk size; loop until we get valid selection
diskvalid="no"
while [[ $diskvalid = "no" ]]; do
    read -p "${white}Please input desired zLinux disk size:  (3GB, 9GB, 27GB): ${reset} " dsize

    case "$dsize" in
    3*)
	# test if there is at least 3GB of disk space
        FREE=`df -k --output=avail "$PWD" | tail -n1`
        if [[ $FREE -lt 4545728 ]]; then               # 3G = 3*1024*1024k + ISO + copy_of_ISO
           echo "${rev}${red}Sorry you don't have enough disk space. Terminating now... ${reset}"   # less than 10GBs free!
           exit 1
        fi
        echo "${yellow}Roger, ${cyan} 3GB  ${reset}"
        logit "user asked for 3GB DASD size"
        [ -e ./dasd/hd0.120 ] && rm -f dasd/hd0.120 # remove if it exists
        dasdinit64 -z ./dasd/hd0.120 3390-3 HD0 > logs/dasdinit.log 2> ./logs/dasdinit_error.log
        dasdresult=$?
        diskvalid="yes"
        ;;
    9*)
        FREE=`df -k --output=avail "$PWD" | tail -n1` 
        if [[ $FREE -lt 10837184 ]]; then               # 9G = 9*1024*1024k + IS + copy_of_ISO
           echo "${rev}${red}Sorry you don't have enough disk space. Terminating now... ${reset}"   # less than 10GBs free!
           exit 1
        fi
        echo "${yellow}Roger, ${cyan} 9GB ${reset}"
        logit "user asked for 9GB DASD size"
        [ -e ./dasd/hd0.120 ] && rm -f dasd/hd0.120 # remove if file already exists
        dasdinit64 -z ./dasd/hd0.120 3390-9 HD0 > logs/dasdinit.log 2> ./logs/dasdinit_error.log
        dasdresult=$?
        diskvalid="yes"
        ;;
    27*)
        FREE=`df -k --output=avail "$PWD" | tail -n1`  
        if [[ $FREE -lt 29711552 ]]; then               # 27G = 27*1024*1024k + ISO  + copy_of_ISO!!
           echo "${rev}${red}Sorry you don't have enough disk space. Terminating now... ${reset}"   # less than 10GBs free!
           exit 1
        fi
        echo "${yellow}Roger, ${cyan} 27GB ${reset}"
        logit "user asked for 27GB DASD size"
        [ -e ./dasd/hd0.120 ] && rm -f dasd/hd0.120 # remove if file already exists
        dasdinit64 -z ./dasd/hd0.120 3390-27 HD0 > logs/dasdinit.log 2> ./logs/dasdinit_error.log
        dasdresult=$?
        diskvalid="yes"
        ;;
    *)
        echo "${red}Unrecognized selection: $dsize. Try again and supply correct input, either 3GB, 9GB, or 27GB.${reset}"
        ;;
    esac

    if [[ $dasdresult -ne 0 ]]; then
        echo "${red}Error creating DASD file. Check logs/dasdinit.log and logs/dasdinit_error.log.${reset}"
        exit 1
    fi
done

# ask for confirmation before downloading iso....
read -p "${white}Continue with 700MB ISO download? (Y/N): " confirm && [[ $confirm == [yY] || $confirm == [yY][eE][sS] ]] || exit 1
echo "${reset}"
# user said it's ok to download. get iso
./scripts/getiso || exit 1

# Get user password and embed the preseed file into the initrd for install
./scripts/config_preseed

# execute network configurator
$SUDO ./scripts/set_network
$SUDO chown $(id -u):$(id -g) ./logs/setnetwork.log*

# The hercifc util needs to be setuid root to manage network interface. We will
# copy the as-installed copy, which is owned by the user and not root, so that
# we can restore the original owner and permissions after Hercules is done.
# But we won't overwrite an existing backup, in case previous runs failed and
# we're starting again
if [[ ! -f herc4x/bin/hercifc.orig ]]; then
    cp herc4x/bin/hercifc herc4x/bin/hercifc.orig
fi
$SUDO chown root:root herc4x/bin/hercifc
$SUDO chmod +s herc4x/bin/hercifc

# copy correct .rc file for installation
cp templates/hercules.rc.DVD hercules.rc
chmod +w hercules.rc   # ...just in case, for easy replacement later

echo "${yellow}Starting hercules with installation script now. Be patient. ${reset}"
read -p "${white}Please press ENTER to continue with install now. ${reset}" pressenter
# just giving user a chance to see
logdate=`date "+%F-%T"`
FILE=./logs/hercules.log.$logdate
$SUDO HERCULES_RC=hercules.rc hercules -f hercules.cnf > "$FILE"

logit "finished hercules run"
# After hercules finishes, restore the original hercifc
# This is for two reasons: less time leaving a suid binary sitting around,
# and make it easier for the user to clean up later by "rm -rf ..." this
# entire directory and not run into problems deleting a file owned by root.
if [[ -f herc4x/bin/hercifc.orig ]]; then
    $SUDO rm -f herc4x/bin/hercifc
    mv herc4x/bin/hercifc.orig herc4x/bin/hercifc
fi

if [[ ! -f install_success ]]; then
    logit "install determined to be a failure!!!!!"	
    echo "${rev}${red}It seems Hercules quit before the installation finished successfully."
    echo "Check for errors in the logs in logs/, and try running the installation script again.${reset}"
    exit 1
fi

# copy the correct hercules.rc for future use
cp templates/hercules.rc.hd0 hercules.rc


logit "Successful install! Yesh!"
echo "${yellow}It seems that the installation was successful. Start it with: ${reset}"
echo "${magenta}./run_zlinux.bash ${reset}"
echo
echo "${yellow}Good bye!${reset}"

logit "zlinux_install is now quitting"

# moshix LICENSES THE LICENSED SOFTWARE "AS IS," AND MAKES NO EXPRESS OR IMPLIED
# WARRANTY OF ANY KIND. moshix SPECIFICALLY DISCLAIMS ALL INDIRECT OR IMPLIED
# WARRANTIES TO THE FULL EXTENT ALLOWED BY APPLICABLE LAW, INCLUDING WITHOUT
# LIMITATION ALL IMPLIED WARRANTIES OF, NON-INFRINGEMENT, MERCHANTABILITY, TITLE
# OR FITNESS FOR ANY PARTICULAR PURPOSE. NO ORAL OR WRITTEN INFORMATION OR ADVICE
# GIVEN BY moshix, ITS AGENTS OR EMPLOYEES SHALL CREATE A WARRANTY
